src/openfermion/utils/_trotter_exp_to_qgates_test.py

```diff
@@ -229,3 +230,124 @@ def test_qasm_string_XYZ(self):
 Rx -1.5707963267948966 3'''
 
         self.assertEqual(qasmstr, strcorrect)
+
+    def test_qasm_string_Controlled_XYZ(self):
+        # Testing for correct QASM string output w/ Pauli-{X,Y,Z}
+        # QubitOperator('X0 Z1 Y3', 0.5) and a controlled ancilla
+
+        # Number of qubits
+        qasmstr = str(count_qubits(self.opA)+1) + "\n"
+
+        # Write each QASM operation
+        qasmstr += "\n".join(
+            trotterize_exp_qubop_to_qasm(self.opA, ancilla='ancilla'))
+
+        # Correct string
+        strcorrect = '''5
+H 0
+Rx 1.5707963267948966 3
+CNOT 0 1
+CNOT 1 3
+C-Phase 0.5 3 ancilla
+CNOT 1 3
+CNOT 0 1
+H 0
+Rx -1.5707963267948966 3'''
+
+        self.assertEqual(qasmstr, strcorrect)
+
+    def test_qasm_string_Controlled_XYZ_ancilla_list(self):
+                # Testing for correct QASM string output w/ Pauli-{X,Y,Z}
+        # QubitOperator('X0 Z1 Y3', 0.5) and a controlled ancilla
+
+        # Number of qubits
+        qasmstr = str(count_qubits(self.opA)+1) + "\n"
+
+        qubit_list = ['q0', 'q1', 'q2', 'q3']
+
+        # Write each QASM operation
+        qasmstr += "\n".join(
+            trotterize_exp_qubop_to_qasm(self.opA, ancilla='ancilla',
+                                         qubit_list=qubit_list))
+
+        # Correct string
+        strcorrect = '''5
+H q0
+Rx 1.5707963267948966 q3
+CNOT q0 q1
+CNOT q1 q3
+C-Phase 0.5 q3 ancilla
+CNOT q1 q3
+CNOT q0 q1
+H q0
+Rx -1.5707963267948966 q3'''
+
+        self.assertEqual(qasmstr, strcorrect)
+
+    def test_qasm_string_controlled_identity(self):
+        qasmstr = str(count_qubits(self.op_id)+1) + "\n"
+
+        # Write each QASM operation
+        qasmstr += "\n".join(
+            trotterize_exp_qubop_to_qasm(self.op_id, ancilla='ancilla'))
+
+        strcorrect = '''1
+Rz 1.0 ancilla'''
```

Comment:
This test is a little confusing.  Typically if we are controlling on an operation, we hope to see something like
C-op angle control target
but we get an unusual result because it's a controlled identity where the control is the same as the target.  Do we want to allow such a behavior or raise an exception if the control is identical to the target?

Reply:
I think we want to allow this behavior. If we raise an error then no-one can pass in a Hamiltonian with any coefficient in front of the identity (before this commit that raised an error). And if we just ignore the identity term then every time you did phase estimation with the circuit you'd wind up with a set of phases offset from their true variables (which I can't imagine someone would want).

Definitely the gate is the right choice, although I agree that the nomenclature is a bit odd - I guess technically this is a controlled global phase on the system, but that is just a Rz rotation, and I'm not sure if people would understand what the other gate meant when they read the output.

But, this again raises a good question - does OpenFermion have a set of standards for the gates that it outputs? If not, it might be a good idea to write this for future reference (and so that people that want to use the program don't have to dig into the code to find out what it might give them).

Reply:
The code that you are editing was never designed as a particularly serious way of compiling circuits in OpenFermion. In fact, the core OpenFermion library is currently not really designed to compile things all the way to circuits - it is more of a set of tools for obtaining and manipulating representations of Hamiltonians and states of relevance to quantum simulations of electronic structure. The idea was always to compile circuits using plugins that link to specific quantum programming frameworks. For instance, there is Forest-OpenFermion to use the Rigetti stack and there is OpenFermion-ProjectQ to use the ProjectQ stack. Google is also working a similar framework for our chips and we'll also release an OpenFermion plugin designed to work with that. We also welcome the creation of new plugins like this.

So the answer to your question "does OpenFermion have a set of standards for the gates that it outputs?" is "no" because we never really intended OpenFermion core to a serious tool for compiling all the way to the hardware level. That said, I think it is fine to add the code you are adding.

Reply:
Fair point about the action of a controlled global phase, whatever that might mean.